contracts/fatex/v3/FateRewardControllerV3.sol

```diff
@@ -0,0 +1,536 @@
+
+        if (fateReward > 0) {
+            fate.transferFrom(vault, address(this), fateReward);
+            pool.accumulatedFatePerShare = pool.accumulatedFatePerShare
+                .add(fateReward.mul(1e12).div(lpSupply));
+        }
+        pool.lastRewardBlock = block.number;
+    }
+
+    // Deposit LP tokens to MasterChef for FATE allocation.
+    function deposit(uint256 _pid, uint256 _amount) public {
+        PoolInfo storage pool = poolInfo[_pid];
+        IFateRewardController.UserInfo memory user = _getUserInfo(_pid, msg.sender);
+        updatePool(_pid);
+        if (user.amount > 0) {
+            _claimRewards(_pid, msg.sender, user, pool);
+        }
+        pool.lpToken.safeTransferFrom(
+            address(msg.sender),
+            address(this),
+            _amount
+        );
+
+        uint userBalance = user.amount.add(_amount);
+        _userInfo[_pid][msg.sender] = UserInfoV2({
+            amount : userBalance,
+            rewardDebt : userBalance.mul(pool.accumulatedFatePerShare).div(1e12),
+            isUpdated : true
+        });
+        _recordDepositBlock(_pid, msg.sender);
+        emit Deposit(msg.sender, _pid, _amount);
+    }
+
+    // Withdraw LP tokens from MasterChef.
+    function withdraw(uint256 _pid, uint256 _amount) public {
+        PoolInfo storage pool = poolInfo[_pid];
+        IFateRewardController.UserInfo memory user = _getUserInfo(_pid, msg.sender);
+        require(user.amount >= _amount, "withdraw: not good");
+        updatePool(_pid);
+
+        _claimRewards(_pid, msg.sender, user, pool);
+
+        uint userBalance = user.amount.sub(_amount);
+        _userInfo[_pid][msg.sender] = UserInfoV2({
+            amount : userBalance,
+            rewardDebt : userBalance.mul(pool.accumulatedFatePerShare).div(1e12),
+            isUpdated : true
+        });
+
+        uint256 withdrawAmount = _amount;
+        if (isFatePool[_pid]) {
+            withdrawAmount = _reduceFeeAndUpdateMembershipInfo(
+                _pid,
+                msg.sender,
+                withdrawAmount,
+                _amount == user.amount
+            );
+        }
+
+        pool.lpToken.safeTransfer(msg.sender, withdrawAmount);
+        emit Withdraw(msg.sender, _pid, withdrawAmount);
+    }
+
+    // reduce lpWithdrawFee and lockedRewardFees
+    // if withdraw all, add current reward points to additional user points and do not earn any new user points
+    function _reduceFeeAndUpdateMembershipInfo(
+        uint256 _pid,
+        address _account,
+        uint256 _amount,
+        bool _withdrawAll
+    ) internal returns(uint256 withdrawAmount) {
+        // minus LockedRewardFee
+        userLockedRewards[_pid][_account] = userLockedRewards[_pid][_account]
+            * (1e18 - _getLockedRewardsFeePercent(_pid, _account))
+            / 1e18;
+
+        // minus LPWithdrawFee
+        withdrawAmount = _amount
+            * (1e18 - _getLPWithdrawFeePercent(_pid, _account))
+            / 1e18;
```

Comment:
good point. since i used higher solidity version before (> 8.0), i used direct operators.
i forgot this is an older solidity version, will fix them soon.